src/ms3/bs4_parser.py

```diff
@@ -3169,6 +3312,101 @@ def set_first_timesig(self, first_timesig: str):
                 _ = self.new_tag("sigN", value=sigN, append_within=timesig_tag)
                 _ = self.new_tag("sigD", value=sigD, append_within=timesig_tag)
 
+    def set_first_tempo(self, active_tempo_tag: bs4.Tag):
+        self.enforce_tempo(piece_tempo_tag=active_tempo_tag, user_call=False)
+
+    def replace_notes_before_onset(
+        self,
+        start_mc: int,
+        mc_onset: Fraction,
+    ):
+        staves = self.tags[start_mc]
+        for staff, voices in staves.items():
+            for voice, onsets in voices.items():
+                for onset, tag_dicts in onsets.items():
+                    if onset >= mc_onset:
+                        continue
+                    for tag_dict in tag_dicts:
+                        if tag_dict["name"] != "Chord":
+                            continue
+                        from_chord_to_rest(tag_dict["tag"])
+
+    def replace_notes_after_onset(
+        self,
+        end_mc: int,
+        mc_onset: Fraction,
+        exclude_end: bool = False,
+    ):
+        staves = self.tags[end_mc]
+        for staff, voices in staves.items():
+            for voice, onsets in voices.items():
+                for onset, tag_dicts in onsets.items():
+                    if onset == mc_onset and not exclude_end:
+                        continue
+                    elif onset < mc_onset:
+                        continue
+                    for tag_dict in tag_dicts:
+                        if tag_dict["name"] != "Chord":
+                            continue
+                        from_chord_to_rest(tag_dict["tag"])
+
+    def enforce_tempo(
+        self,
+        piece_tempo_tag: Optional[bs4.Tag] = None,
+        user_tempo: Optional[float] = None,
+        user_beat_factor: Optional[Fraction] = Fraction(1 / 1),
+        user_call: Optional[bool] = True,
+    ):
+        for measure_tag in self.iter_first_measures():
+            tempo_tag = measure_tag.find("Tempo")
+            timesig_tag = measure_tag.find("TimeSig")
+            if not user_call and piece_tempo_tag is not None and tempo_tag is None:
+                # Copying active tempo tag from "parent" piece
+                timesig_tag.insert_after(piece_tempo_tag)
+                return
```

Comment:
I'm using the `return` statement in this case because I only want to do the operation on the first measure of the execrpt but since I'm iterating over all of them, I'm using the `return` to stop the loop. I'm guessing there is a way to only take the first measure without needing to iterate through them all...?